pubtools/pulplib/_impl/client/client.py

```diff
@@ -1054,3 +1054,46 @@ def _do_sync(self, repo_id, sync_options):
         return self._task_executor.submit(
             self._do_request, method="POST", url=url, json=body
         )
+
+    def create_repository(self, repo):
+        """Create a repository with initial data provided in the
+           argument. Importer for repository is automatically associated
+           if available. If repository already exists, only warning is logged.
+
+        Args:
+            repo (:class:`~pubtools.pulplib.Repository`)
+                A repository object used for creation.
+
+        Returns:
+            Future
+                A future which is resolved with a value of ``None`` once the
+                repository has been created.
+
+        .. versionadded:: 2.39.0
+        """
+        url = os.path.join(self._url, "pulp/api/v2/repositories/")
+
+        body = repo._to_data()
+        repo_id = body["id"]
+
+        importer = body.pop("importers", [])
+        body["importer_type_id"] = importer[0]["importer_type_id"] if importer else None
+        body["importer_config"] = importer[0]["config"] if importer else None
+
+        def log_existing_repo(exception):
+            if (
+                getattr(exception, "response", None) is not None
+                and exception.response.status_code == 409
+            ):
+                LOG.warning("Repository %s already exists", repo_id)
```

Comment:
I feel a bit uncomfortable that we swallow these errors completely.  It seems like a trap, because someone can do this:

- write a script which creates a repo accidentally setting the wrong config, run it against a server
- realize their mistake, fix the config
- run it again
- their script "succeeds" but it output this warning and they didn't notice.  They believe their script is working OK but actually the importer config is permanently incorrect.

I understand that some handling of 409 would be needed to make the API able to be retried.

What do you think about extending this API so it'd check the repo before returning?
I mean keeping this 409 handling as it is, but at the end of the method it could do:

- fetch the newly created repo
- compare against the one passed in by the user
- if the user's passed Repository set "importer", then compare against the importer actually on the repo, and raise an error if they differ
- the fetched Repository object can be returned (return type `Future[Repository]`)

I believe something like that might be a good idea anyway, because the `Repository` object that the user passed in here is not connected to the client so most of the methods on it can't be used.  The client in most cases probably wants to do something with the `Repository` object, so if this API doesn't fetch and return it they'll probably have to do it themselves afterward.  If they're going to do it anyway, we may as well do it for them which gives us a chance to check if the repo is really what was requested.